README.md
# Personode

Personode is a MATLAB toolbox designed for regions-of-interest (ROI) individualization based on fMRI spatial independent component analysis (ICA) maps after a semi-automatic classification into resting-state networks (RSNs).

## Requirements

 - MATLAB R2013A or newer
 - SPM12 (get at https://www.fil.ion.ucl.ac.uk/spm/software/spm12/)
 - MarsBaR ('marsbar-0.44.zip' is included. Alternatively, get it at http://marsbar.sourceforge.net/)

## A Quick Intro

 - Download the compressed file (get at https://www.nitrc.org/projects/personode) and extract it to a folder of your preference or clone this repository to your desktop.
 - Set Personode's path to MATLAB. Make sure that both SPM12 and MarsBaR paths are also included (do not include "spm2", "spm5", and "spm99" contained in MarsBaR folder).
 - Type Personode in Matlab's command window. The following GUI will show up:

![Personode GUI in Windows](Personode_Picture_1.png)

 - In "Source Files Main Options":
	- Specify the number of subjects;
	- In "Input files mode", if "Automatic" is selected, only the ICA file from the first subject will be required. If "Manual" is selected, one needs to manually select the desired files for analysis;
	- There are two modes of ROI definition: spherical ROIs (radius can be altered) and probabilistic irregular ROIs. Select one of them;
	- Specify the output folder.
- In "Defining Paths", ICA map files must be loaded. GIFT toolbox (download at http://mialab.mrn.org/software/gift/) for MATLAB can be used to create the components, for example.
- In "Templates Definition", choose the resting-state network(s) you would like to identify.
- In "Are ICA maps coregistered", select 'yes' if the images already have the MNI dimensions. Select 'no' if you want the toolbox to perform coregistration for you (in this case, you need to further specify an MNI-normalized anatomical file in a forthcoming SPM window).
	- If the images are already in the MNI dimensions, you can navigate through the components using the GUI buttons (slice, threshold, subject, component, refresh image) 
- In "Describing ROI features", select the information you would like to save from the ROI definition process.
- In "How will the nodes be created", select either "Separated" for creating a mask with the regions separately labeled as integer numbers or "As a network" as a binary mask for networks.
- Press Start.
- A Manual selection window will appear showing, in the left side, the RSN template you want to identify and the three most probable maps to be classified. You can navigate using the xyz buttons. Probability values are also shown in red to help in the classification process. One can choose a map and then press Next. If more than one RSN was selected to be identified, one can also skip the current map to be defined later.
- Nifti masks and TXT info files will be saved in the specified output folder.

## Example using coregistered images

 - We will begin this example using the images included in the folder 'Personode\Images_example\Coregistered images'
 - 'rica_mean_coment_ica_s_all_.nii' is the resulting file of a group ICA with 20 components.
 - 'rica_sub001_component_ica_s1_.nii' and 'rica_sub002_component_ica_s1_.nii' are the resulting subject-level maps for two subjects.
 - Thus, set '# of Subjects' to 2, set your preferred output folder and options. An example is shown below.

 ![Source Files Main Options](Personode_Picture_2_a.png)

 - Then fill in the 'ICA Group File' and 'Individual ICA Files' paths. If you previously set 'Automatic` as the 'input file mode', you only need to select the ICA map for the first subject.
 - Select the networks of interest in 'Templates Definition'. 

![Templates Definition](Personode_Picture_2_c.png)

 - Since the ICA maps are already in the MNI dimensions, select yes.

![Are ICA Maps coregistered](Personode_Picture_2_d.png)

 - After selecting this, the image interface will be unlocked. You can freely explore either group or subjects' maps.

![Spatial components](Personode_Picture_2_e.png)

 - Finally, select your 'Describing ROIs features' of interest, and the way you want the ROIs to be created. Here you have two options: 'Separated' or 'As a network'. Filling this information will unlock the 'Start' button.

![Spatial components](Personode_Picture_2_f.png) 

 - Clicking this button will show this message on MATLAB's command window.

![Start message](Personode_Picture_2_g.png) 

 - Shortly after, you will be asked to selected maps to be classified as RSNs. You can skip a network at any time, later you will be asked again about the skipped ones.

 ![Start message](Personode_Picture_2_h.png) 

 - Personode will then create the ROIs for the group and each subject.

 ![Positioning nodes](Personode_Picture_2_i.png) 

 - In the selected output folder, ROIs masks will be created for all inputs.

![Rois masks](Personode_Picture_2_j.png) 

 - A 'ROIs_Description_*.txt' file is also created for each input. An example is shown below.

![Rois description](Personode_Picture_2_k.png) 


# References

 - \[MarsBaR\] Matthew Brett, Jean-Luc Anton, Romain Valabregue, Jean-Baptiste Poline. _Region of interest analysis using an SPM toolbox_ \[abstract\] Presented at the 8th International Conference on Functional Mapping of the Human Brain, June 2-6, 2002, Sendai, Japan. Available on CD-ROM in NeuroImage, Vol 16, No 2.
 - \[SPM12\] Penny, W. D., Friston, K. J., Ashburner, J. T., Kiebel, S. J., & Nichols, T. E. (Eds.). (2007). Statistical parametric mapping: the analysis of functional brain images.
 


NEWS.md
# v0.2

Implemented quite a few fixes to Personode GUI. Not limited to:

 - Using radio buttons where warranted instead of checkboxes
 - Fixing some instances where not selecting images or other parameters or returning empty selections would freeze or glitch the GUI
 - Added requirements for both SPM12 and MarsBar; bundling 'marsbar-0.44.zip' (GNU GPL licensed) with the library
 - Added a step-by-step example  (with pictures!) and also a brief introduction to the toolbox
 - Hiding the GUI when it's not interactive

findPeaks.m
function [peak_value,vec,x,y,z,MNIx,MNIy,MNIz,atlasInfo,exc,mask]=findPeaks(tmp_names,compImage,class,subj,vec,exc,sep,workPath)

D = dir([workPath filesep 'Clusters']);
atlasInfo=spm_vol([workPath filesep 'Clusters' filesep D(3).name]);

if subj==1
    
    k=1;
    X=char(D(3:end,1).name);
    Y=str2num(X(:,4:5));
    for i=1:length(tmp_names)
        n_selClusters(i)=nnz(Y==tmp_names(i));
    end
    
    vec=[];
    class_without_zeros=class;
    class_without_zeros(class_without_zeros==0)=[];
    for i=1:length(tmp_names)
        vec=[vec class_without_zeros(i)*ones(1,n_selClusters(i))];
    end
    
end

mask=zeros(91,109,91,length(tmp_names));
n=1;
m=1;
if subj==1
    k=1;
    for i=3:length(D)
        if nnz(str2num(D(i).name(4:5))==tmp_names)
            if ~nnz(k==exc)
                VallROIs=spm_vol([workPath filesep 'Clusters' filesep D(i).name]);
                Clusters=spm_read_vols(VallROIs); % read nodes templates
                zica=compImage(:,:,:,vec(k));
                zica=Clusters.*zica;
                
                % maximum z value inside the cluster
                idx2=find(Clusters>.5);
                
                [peak_value(k) idx3(k)]=max(zica(idx2));
                [x(k) y(k) z(k)]=ind2sub(size(zica),idx2(idx3(k)));
                x(k)=91-x(k)+1;
                [MNIx(k) MNIy(k) MNIz(k)] = mni2orFROMxyz_noprint(x(k),y(k),z(k),2,'xyz');
                
                [firstValue,idx3]=sort(zica(idx2),'descend');
                
                n_vox_total=nnz(Clusters);
                if n_vox_total/10<120
                    n_vox=120;
                else
                    n_vox=n_vox_total/10;
                end
                
            end
            
            if k~=1
                if vec(k)~=vec(k-1)
                    n=n+1;
                    m=1;
                end
            end
            
            if peak_value(k)>=0.1
                if sep == 1
                    im=m*(zica>=firstValue(round(n_vox)));
                    mask(:,:,:,n)=mask(:,:,:,n)+im;
                    m=m+1;
                else
                    im=zica>=firstValue(round(n_vox));
                    mask(:,:,:,n)=mask(:,:,:,n)+im;
                end
            end
            k=k+1;
        end
    end
else
    k=1;
    for i=3:length(D)
        if nnz(str2num(D(i).name(4:5))==tmp_names)
            D2(i-2,:)=D(i).name;
        end
    end
    
    j=1;
    for i=1:size(D2,1)
        if D2(i,1)=='R'
            D3(j,:)=D2(i,:);
            j=j+1;
        end
    end
    
    D3(exc,:)=[];
    
    for i=1:size(D3,1)
        VallROIs=spm_vol([workPath filesep 'Clusters' filesep D3(i,:)]);
        Clusters=spm_read_vols(VallROIs); % read nodes templates
        zica=compImage(:,:,:,vec(i));
        zica=Clusters.*zica;
        
        idx2=find(Clusters>.5);
        
        [peak_value(i) idx3(i)]=max(zica(idx2));
        [x(i) y(i) z(i)]=ind2sub(size(zica),idx2(idx3(i)));
        x(i)=91-x(i)+1;
        [MNIx(i) MNIy(i) MNIz(i)] = mni2orFROMxyz_noprint(x(i),y(i),z(i),2,'xyz');
        
        [firstValue,idx3]=sort(zica(idx2),'descend');
        
        n_vox_total=nnz(Clusters);
        if n_vox_total/10<120
            n_vox=120;
        else
            n_vox=n_vox_total/10;
        end
        
        if k~=1
            if vec(k)~=vec(k-1)
                n=n+1;
                m=1;
            end
        end
        
        if peak_value(k)>=0.1
            if sep == 1
                im=m*(zica>=firstValue(round(n_vox)));
                mask(:,:,:,n)=mask(:,:,:,n)+im;
                m=m+1;
            else
                im=zica>=firstValue(round(n_vox));
                mask(:,:,:,n)=mask(:,:,:,n)+im;
            end
        end
        k=k+1;
    end
    
end

if subj==1
    exc=find(peak_value<0.1); % eliminate points when low peaks are found
    peak_value(exc)=[];
    x(exc)=[];
    y(exc)=[];
    z(exc)=[];
    MNIx(exc)=[];
    MNIy(exc)=[];
    MNIz(exc)=[];
    vec(exc)=[];
end

mask=flipud(mask);
    
return

mni2orFROMxyz_noprint.m
%% Convert MNI coords to Voxel XYZ (3D matrix) coordinates
%Written by Michael T Rubens, June 4 2008, UC San Francisco, Gazzaley Lab
function [x y z outtype] = mni2orFROMxyz_noprint(x, y, z,vs,coordef)
%x  = x coord to convert, OR, 1D index (voxel position, i.e., from find(n))
%y  = y coord to convert, OR, dimensions of matrix (i.e., [91 109 91])
%z  = z coord to convert
%vs = voxel size in mm (default is 2mm)
%       pass in empty z (i.e., z=[]) to use 1D index and custom voxel size
%
% coordef = 'mini' or 'xyz' (string, case-insensitive). Type of input coordinates. (default is 'xyz')
%       pass in empty vs (i.e., vs[]) to use default voxel size and still specify coordinate type

if nargin<3 || isempty(z)
    [x y z] = ind2sub(y,x);
end

if nargin<4 || isempty(vs)
    vs = 2;%2mm
end

if nargin<5
    coordef = 'mni';
end

% MNI origin in voxel coordinates (Anterior Commisure)
% origin = 352852; (origin index)
origin = [46 64 37]; % [X Y Z]

switch lower(coordef)
        case 'mni'
%             x=origin(1)- x/vs;
            x=x/vs + origin(1);
            y=y/vs + origin(2);
            z=z/vs + origin(3);
            
            outtype = 'XYZ';
    case 'xyz'
%             x = (origin(1)-x)*vs;
            x = (x-origin(1))*vs;
            y = (y-origin(2))*vs;
            z = (z-origin(3))*vs;
            
            outtype = 'MNI';
    otherwise
        error('unknown coordinate definition. Input EITHER ''mni'' or ''xyz''');
end

% fprintf('Outputting %s coordinates\n',outtype);

end

manual_selection.m
function S=manual_selection(temp,img1,img2,img3,sel_prob,sel_temp,idx2,idx3)

S = struct('First',1);
S.str=[45 55 45];
tol=.01;
S.firstPrint=1;
S.skip=0;
S.idx2=idx2;
S.idx3=idx3;
S.select=[0 0 0];

network_names={'Dorsal Attention','Auditory','Basal Ganglia','Cerebellum','Cognition/Emotion',...
    'Default Mode, Dorsal/Posterior','Default Mode, Medial','Executive Control','Frontoparietal, Left',...
    'Frontoparietal, Right','Ventral Stream, Left','Ventral Stream, Right','Postcentral','Precentral',...
    'Precuneus','Salience','Sensorimotor','Visual, Primary I','Visual, Primary II','Visual, Medial',...
    'Visual, Higher Level'};

handles_gui.h = figure;
scr_dim=get(0, 'ScreenSize');

set(handles_gui.h, ...
    'position', scr_dim, ...
    'name', 'Manual selection', ...
    'tag', 'h');

uicontrol('style', 'pushbutton', ...
    'Parent',handles_gui.h,...
    'units', 'normalized',...
    'position', [.85 .05 .05 .035], ...
    'FontSize', 12*scr_dim(4)/1080,...
    'callback', {@skip_press},...
    'value',0,...
    'string', 'Skip');
uicontrol('style', 'pushbutton', ...
    'Parent',handles_gui.h,...
    'units', 'normalized',...
    'position', [.9 .05 .05 .035], ...
    'FontSize', 12*scr_dim(4)/1080,...
    'FontWeight', 'bold',...
    'Enable', 'off',...
    'Tag','next',...
    'callback', {@next_press},...
    'string', 'Next');

S.tmp=temp;
S.tmp(S.tmp==0)=nan;

S.img1=img1;
S.img1(abs(S.img1)<tol)=nan;

S.img2=img2;
S.img2(abs(S.img2)<tol)=nan;

S.img3=img3;
S.img3(abs(S.img3)<tol)=nan;

hp1 = uipanel('Title','Template',...
    'FontSize',12*scr_dim(4)/1080,...
    'BackgroundColor',[.85 .85 .85],...
    'Position',[.05 .1 .2 .8],...
    'Tag','TmpPanel',...
    'Clipping','on');
uicontrol('style', 'text', ...
    'Parent',hp1,...
    'BackgroundColor', [0.85 0.85 0.85],...
    'units', 'normalized',...
    'position', [.02 .02 .96 .05], ...
    'FontSize', 11*scr_dim(4)/1080,...
    'FontWeight', 'bold',...
    'ForegroundColor', 'blue',...
    'string', ['RSN to be defined: ' char(network_names(sel_temp))]);

hp2 = uipanel('Title','Component #1',...
    'FontSize',12*scr_dim(4)/1080,...
    'BackgroundColor',[.85 .85 .85],...
    'Position',[.31 .1 .2 .8],...
    'Tag','img1Panel',...
    'Clipping','on');
A0 = axes;
set(A0,'Parent',hp2,...
    'NextPlot', 'replacechildren');
uicontrol('style', 'text', ...
    'Parent',hp2,...
    'BackgroundColor', [0.85 0.85 0.85],...
    'units', 'normalized',...
    'position', [.02 .02 .96 .05], ...
    'FontSize', 13*scr_dim(4)/1080,...
    'FontWeight', 'bold',...
    'ForegroundColor', 'red',...
    'tag', 'prob1',...
    'string', ['probability = ' num2str(sel_prob(1), '%01.f') '%']);
uicontrol('style', 'pushbutton', ...
    'Parent',hp2,...
    'units', 'normalized',...
    'position', [.4 .95 .2 .03], ...
    'FontSize', 12*scr_dim(4)/1080,...
    'FontWeight', 'bold',...
    'callback', {@select1_press},...
    'tag', 'select1',...
    'string', 'Select');

if S.idx2~=0

hp3 = uipanel('Title','Component #2',...
    'FontSize',12*scr_dim(4)/1080,...
    'BackgroundColor',[.85 .85 .85],...
    'Position',[.53 .1 .2 .8],...
    'Tag','img2Panel',...
    'Clipping','on');
A0 = axes;
set(A0,'Parent',hp3,...
    'NextPlot', 'replacechildren');
uicontrol('style', 'text', ...
    'Parent',hp3,...
    'BackgroundColor', [0.85 0.85 0.85],...
    'units', 'normalized',...
    'position', [.02 .02 .96 .05], ...
    'FontSize', 13*scr_dim(4)/1080,...
    'FontWeight', 'bold',...
    'ForegroundColor', 'red',...
    'tag', 'prob2',...
    'string', ['probability = ' num2str(sel_prob(2), '%01.f') '%']);
uicontrol('style', 'pushbutton', ...
    'Parent',hp3,...
    'units', 'normalized',...
    'position', [.4 .95 .2 .03], ...
    'FontSize', 12*scr_dim(4)/1080,...
    'FontWeight', 'bold',...
    'callback', {@select2_press},...
    'tag', 'select2',...
    'string', 'Select');

end

if S.idx3~=0
    
    hp4 = uipanel('Title','Component #3',...
        'FontSize',12*scr_dim(4)/1080,...
        'BackgroundColor',[.85 .85 .85],...
        'Position',[.75 .1 .2 .8],...
        'Tag','img3Panel',...
        'Clipping','on');
    A0 = axes;
    set(A0,'Parent',hp4,...
        'NextPlot', 'replacechildren');
    uicontrol('style', 'text', ...
        'Parent',hp4,...
        'BackgroundColor', [0.85 0.85 0.85],...
        'units', 'normalized',...
        'position', [.02 .02 .96 .05], ...
        'FontSize', 13*scr_dim(4)/1080,...
        'FontWeight', 'bold',...
        'ForegroundColor', 'red',...
        'tag', 'prob3',...
        'string', ['probability = ' num2str(sel_prob(3), '%01.f') '%']);
    uicontrol('style', 'pushbutton', ...
        'Parent',hp4,...
        'units', 'normalized',...
        'position', [.4 .95 .2 .03], ...
        'FontSize', 12*scr_dim(4)/1080,...
        'FontWeight', 'bold',...
        'callback', {@select3_press},...
        'tag', 'select3',...
        'string', 'Select');
    
end

% Select x
uicontrol('style', 'text', ...
    'Parent',handles_gui.h,...
    'BackgroundColor', [0.9 0.9 0.9],...
    'units', 'normalized',...
    'position', [.78 .95 .02 .02], ...
    'string', 'x');
uicontrol('style', 'edit', ...
    'Parent',handles_gui.h,...
    'units', 'normalized',...
    'position', [.8 .95 .02 .02], ...
    'callback', {@sliceSelect},...
    'BackgroundColor', 'white',...
    'String', num2str(S.str(1)),...
    'tag', 'SliceX');
uicontrol('style', 'slider', ...
    'Parent',handles_gui.h,...
    'units', 'normalized',...
    'position', [.82 .95 .01 .02], ...
    'callback', {@btnSlider},...
    'Selected', 'off',...
    'Max', 91,...
    'Min', 1, ...
    'Value', 45, ...
    'SliderStep', [1/90,1/9],...
    'tag', 'SliderX');

% Select y
uicontrol('style', 'text', ...
    'Parent',handles_gui.h,...
    'BackgroundColor', [0.9 0.9 0.9],...
    'units', 'normalized',...
    'position', [.84 .95 .02 .02], ...
    'string', 'y');
uicontrol('style', 'edit', ...
    'Parent',handles_gui.h,...
    'units', 'normalized',...
    'position', [.86 .95 .02 .02], ...
    'callback', {@sliceSelect},...
    'BackgroundColor', 'white',...
    'String', num2str(S.str(2)),...
    'tag', 'SliceY');
uicontrol('style', 'slider', ...
    'Parent',handles_gui.h,...
    'units', 'normalized',...
    'position', [.88 .95 .01 .02], ...
    'callback', {@btnSlider},...
    'Selected', 'off',...
    'Max', 109,...
    'Min', 1, ...
    'Value', 55, ...
    'SliderStep', [1/108,1/9],...
    'tag', 'SliderY');

% Select z
uicontrol('style', 'text', ...
    'Parent',handles_gui.h,...
    'BackgroundColor', [0.9 0.9 0.9],...
    'units', 'normalized',...
    'position', [.9 .95 .02 .02], ...
    'string', 'z');
uicontrol('style', 'edit', ...
    'Parent',handles_gui.h,...
    'units', 'normalized',...
    'position', [.92 .95 .02 .02], ...
    'callback', {@sliceSelect},...
    'BackgroundColor', 'white',...
    'String', num2str(S.str(3)),...
    'tag', 'SliceZ');
uicontrol('style', 'slider', ...
    'Parent',handles_gui.h,...
    'units', 'normalized',...
    'position', [.94 .95 .01 .02], ...
    'callback', {@btnSlider},...
    'Selected', 'off',...
    'Max', 91,...
    'Min', 1, ...
    'Value', 45, ...
    'SliderStep', [1/90,1/9],...
    'tag', 'SliderZ');

guidata(handles_gui.h, S);
print_image(S,S.idx2,S.idx3)

uiwait(handles_gui.h);
S = guidata(handles_gui.h);

close(handles_gui.h)

return

function select1_press(src, ~)
handles_gui = guihandles(src);
S = guidata(src);

if S.select(1)==0
    S.select=[1 0 0];
    
    set(handles_gui.select1, 'Selected', 'on');
    set(handles_gui.select2, 'Selected', 'off');
    set(handles_gui.select3, 'Selected', 'off');
    
    set(handles_gui.img1Panel, 'BackgroundColor', [.92 .85 .85]);
    set(handles_gui.img2Panel, 'BackgroundColor', [.85 .85 .85]);
    set(handles_gui.img3Panel, 'BackgroundColor', [.85 .85 .85]);
    
    set(handles_gui.prob1, 'BackgroundColor', [.92 .85 .85]);
    set(handles_gui.prob2, 'BackgroundColor', [.85 .85 .85]);
    set(handles_gui.prob3, 'BackgroundColor', [.85 .85 .85]);
    
    set(handles_gui.next, 'enable', 'on');
else
    S.select=[0 0 0];
    set(handles_gui.select1, 'Selected', 'off');
    set(handles_gui.img1Panel, 'BackgroundColor', [.85 .85 .85]);
    set(handles_gui.next, 'enable', 'off');
end

guidata(handles_gui.h, S);

return

function select2_press(src, ~)
handles_gui = guihandles(src);
S = guidata(src);

if S.select(2)==0
    S.select=[0 1 0];
    
    set(handles_gui.select1, 'Selected', 'off');
    set(handles_gui.select2, 'Selected', 'on');
    set(handles_gui.select3, 'Selected', 'off');
    
    set(handles_gui.img1Panel, 'BackgroundColor', [.85 .85 .85]);
    set(handles_gui.img2Panel, 'BackgroundColor', [.92 .85 .85]);
    set(handles_gui.img3Panel, 'BackgroundColor', [.85 .85 .85]);
    
    set(handles_gui.prob1, 'BackgroundColor', [.85 .85 .85]);
    set(handles_gui.prob2, 'BackgroundColor', [.92 .85 .85]);
    set(handles_gui.prob3, 'BackgroundColor', [.85 .85 .85]);
    
    set(handles_gui.next, 'enable', 'on');
else
    S.select=[0 0 0];
    set(handles_gui.select2, 'Selected', 'off');
    set(handles_gui.img2Panel, 'BackgroundColor', [.85 .85 .85]);
    set(handles_gui.next, 'enable', 'off');
end

guidata(handles_gui.h, S);

return

function select3_press(src, ~)
handles_gui = guihandles(src);
S = guidata(src);

if S.select(3)==0
    S.select=[0 0 1];
    
    set(handles_gui.select1, 'Selected', 'off');
    set(handles_gui.select2, 'Selected', 'off');
    set(handles_gui.select3, 'Selected', 'on');
    
    set(handles_gui.img1Panel, 'BackgroundColor', [.85 .85 .85]);
    set(handles_gui.img2Panel, 'BackgroundColor', [.85 .85 .85]);
    set(handles_gui.img3Panel, 'BackgroundColor', [.92 .85 .85]);
    
    set(handles_gui.prob1, 'BackgroundColor', [.85 .85 .85]);
    set(handles_gui.prob2, 'BackgroundColor', [.85 .85 .85]);
    set(handles_gui.prob3, 'BackgroundColor', [.92 .85 .85]);
    
    set(handles_gui.next, 'enable', 'on');
else
    S.select=[0 0 0];
    set(handles_gui.select3, 'Selected', 'off');
    set(handles_gui.img3Panel, 'BackgroundColor', [.85 .85 .85]);
    set(handles_gui.next, 'enable', 'off');
end

guidata(handles_gui.h, S);

return

function next_press(src, ~)
handles_gui = guihandles(src);
S = guidata(src);

guidata(handles_gui.h, S);
uiresume(handles_gui.h);
return

function skip_press(src, ~)
handles_gui = guihandles(src);
S = guidata(src);

S.skip=1;

guidata(handles_gui.h, S);
uiresume(handles_gui.h);
return

function btnSlider(src, ~)
handles = guihandles(src);
S = guidata(src);

coord = round(get(src,'Value'));

if strcmp(get(src,'Tag'),'SliderX')
    set(handles.SliceX, 'String', coord);
elseif strcmp(get(src,'Tag'),'SliderY')
    set(handles.SliceY, 'String', coord);
else
    set(handles.SliceZ, 'String', coord);
end

set(src,'Value', coord);
guidata(handles.h,S);
print_image(S,S.idx2,S.idx3)

return

function sliceSelect(src, ~)
handles = guihandles(src);
S = guidata(src);

slice = str2double(get(src, 'string'));

if slice<1
    slice=1;
end

if strcmp(get(src,'Tag'),'SliceX')
    if slice>get(handles.SliderX, 'Max')
        slice=get(handles.SliderX,'Max');
    end
    set(handles.SliderX, 'Value', slice);
elseif strcmp(get(src,'Tag'),'SliceY')
    if slice>get(handles.SliderY, 'Max')
        slice=get(handles.SliderY,'Max');
    end
    set(handles.SliderY, 'Value', slice);
else
    if slice>get(handles.SliderZ, 'Max')
        slice=get(handles.SliderZ,'Max');
    end
    set(handles.SliderZ, 'Value', slice);
end

set(src,'string', num2str(slice));
set(src,'Value', slice);

guidata(handles.h, S);
print_image(S,S.idx2,S.idx3)

return

function print_image(S,idx2,idx3)
handles = guihandles;

if S.firstPrint==1
    x=S.str(1);
    y=S.str(2);
    z=S.str(3);
else
    x=get(handles.SliderX,'value');
    y=get(handles.SliderY,'value');
    z=get(handles.SliderZ,'value');
end
    
axialtmp=S.tmp(:,:,z);
coronaltmp=S.tmp(:,y,:);
sagittaltmp=S.tmp(x,:,:);
Atmp=(rot90(axialtmp));
Btmp=(rot90(squeeze(coronaltmp)));
Ctmp=(rot90(squeeze(sagittaltmp)));

axial1=S.img1(:,:,z);
coronal1=S.img1(:,y,:);
sagittal1=S.img1(x,:,:);
A1=(rot90(axial1));
B1=(rot90(squeeze(coronal1)));
C1=(rot90(squeeze(sagittal1)));

axial2=S.img2(:,:,z);
coronal2=S.img2(:,y,:);
sagittal2=S.img2(x,:,:);
A2=(rot90(axial2));
B2=(rot90(squeeze(coronal2)));
C2=(rot90(squeeze(sagittal2)));

axial3=S.img3(:,:,z);
coronal3=S.img3(:,y,:);
sagittal3=S.img3(x,:,:);
A3=(rot90(axial3));
B3=(rot90(squeeze(coronal3)));
C3=(rot90(squeeze(sagittal3)));

A0 = axes;
set(A0,'Parent',handles.TmpPanel,'NextPlot', 'replacechildren');

min_tmp=min(min(min(S.tmp)));
max_tmp=max(max(max(S.tmp)));
lim0= mean([min_tmp max_tmp]);
subplot(3,1,3), imshow(Atmp)
colormap(gca, jet(256));caxis([-lim0 lim0]);
subplot(3,1,1), imshow(Btmp)
colormap(gca, jet(256));caxis([-lim0 lim0]);
subplot(3,1,2), imshow(Ctmp)
colormap(gca, jet(256));caxis([-lim0 lim0]);

A0 = axes;
set(A0,'Parent',handles.img1Panel,'NextPlot', 'replacechildren');

min_img1=min(min(min(S.img1)));
max_img1=max(max(max(S.img1)));
lim1= abs(mean([min_img1 max_img1]));
subplot(3,1,3), imshow(A1)
colormap(gca, jet(256));caxis([-lim1 lim1]);
subplot(3,1,1), imshow(B1)
colormap(gca, jet(256));caxis([-lim1 lim1]);
subplot(3,1,2), imshow(C1)
colormap(gca, jet(256));caxis([-lim1 lim1]);

if S.idx2~=0
    
    A0 = axes;
    set(A0,'Parent',handles.img2Panel,'NextPlot', 'replacechildren');
    
    min_img2=min(min(min(S.img2)));
    max_img2=max(max(max(S.img2)));
    lim2= abs(mean([min_img2 max_img2]));
    subplot(3,1,3), imshow(A2)
    colormap(gca, jet(256));caxis([-lim2 lim2]);
    subplot(3,1,1), imshow(B2)
    colormap(gca, jet(256));caxis([-lim2 lim2]);
    subplot(3,1,2), imshow(C2)
    colormap(gca, jet(256));caxis([-lim2 lim2]);
    
end

if S.idx3~=0
    
    A0 = axes;
    set(A0,'Parent',handles.img3Panel,'NextPlot', 'replacechildren');
    
    min_img3=min(min(min(S.img3)));
    max_img3=max(max(max(S.img3)));
    lim3= abs(mean([min_img3 max_img3]));
    subplot(3,1,3), imshow(A3)
    colormap(gca, jet(256));caxis([-lim3 lim3]);
    subplot(3,1,1), imshow(B3)
    colormap(gca, jet(256));caxis([-lim3 lim3]);
    subplot(3,1,2), imshow(C3)
    colormap(gca, jet(256));caxis([-lim3 lim3]);
    
end

S.firstPrint=0;

guidata(handles.h, S);

return